manage/config.go

```diff
@@ -0,0 +1,126 @@
+package manage
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+)
+
+type ConfigManager struct{}
+
+func NewConfigManager() *ConfigManager {
+	return &ConfigManager{}
+}
+
+func isValidURL(rawURL string) bool {
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return false
+	}
+
+	// Check if the scheme is HTTP or HTTPS
+	if u.Scheme != "http" && u.Scheme != "https" {
+		return false
+	}
+
+	// Check if the host is present
+	if u.Host == "" {
+		return false
+	}
+
+	return true
+}
+
+func (c *ConfigManager) fetch(rawURL string) ([]byte, error) {
+	if !isValidURL(rawURL) {
+		return nil, fmt.Errorf("invalid URL")
+	}
+
+	resp, err := http.Get(rawURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to download config from URL: %v", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return nil, fmt.Errorf("status code:%d returned from remote URL", resp.StatusCode)
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %v", err)
+	}
+
+	if len(body) == 0 {
+		return nil, fmt.Errorf("empty response body from URL")
+	}
+
+	return body, nil
+}
+
+func (c *ConfigManager) writeToDisk(content []byte, filePath string) (int, error) {
+	// create directory path if not already created
+	dirPath := filepath.Dir(filePath)
+	if err := os.MkdirAll(dirPath, 0700); err != nil {
+		return 0, fmt.Errorf("failed to create directory ~/.config/gitleaks : %v", err)
+	}
+
+	// Create a temporary file in the same directory
+	tempFile, err := os.CreateTemp(dirPath, "gitleaks-config-*.tmp")
+	if err != nil {
+		return 0, fmt.Errorf("failed to create temporary file: %v", err)
+	}
+	tempFilePath := tempFile.Name()
+	defer func() {
+		if removeErr := os.Remove(tempFilePath); removeErr != nil {
+			err = fmt.Errorf("failed to remove temporary file: %v (original error: %v)", removeErr, err)
+		} // Clean up the temp file in case of failure
+	}()
+
+	// Write to the temporary file
+	writer := bufio.NewWriter(tempFile)
+	bytesWritten, err := writer.Write(content)
+	if err != nil {
+		return 0, fmt.Errorf("failed to write to temporary file: %v", err)
+	}
+
+	// flush changes to disk
+	if err = writer.Flush(); err != nil {
+		return 0, fmt.Errorf("failed to flush writer: %v", err)
+	}
+
+	if err = tempFile.Close(); err != nil {
+		return 0, fmt.Errorf("failed to close temporary file: %v", err)
+	}
+
+	// replaces or creates the expected file
+	if err = os.Rename(tempFilePath, filePath); err != nil {
+		return 0, fmt.Errorf("failed to rename temporary file: %v", err)
+	}
+
+	return bytesWritten, nil
+}
+
+func (c *ConfigManager) FetchTo(rawURL, targetPath string) error {
+	var (
+		resp []byte
+		err  error
+	)
+
+	resp, err = c.fetch(rawURL)
+	if err != nil {
+		return fmt.Errorf("could not download config from remote url: %w", err)
+	}
+
```

Comment:
I think it'd be a good idea to validate the config before it's written. 

Here's an example of something similar to validate config:

https://github.com/leaktk/scanner/blob/d1372418a8dfd77fdff22d1e335a2eb956ef1651/pkg/scanner/patterns.go#L157-L191

The reference to the single allowlist is a bit stale in that example, but it helps guard against things like:

- Invalid TOML
- Empty responses
- etc